src/mrinufft/operators/off_resonance.py

```diff
@@ -0,0 +1,330 @@
+"""Off Resonance correction Operator wrapper.
+
+Based on the implementation of Guillaume Daval-Frérot in pysap-mri:
+https://github.com/CEA-COSMIC/pysap-mri/blob/master/mri/operators/fourier/orc_wrapper.py
+"""
+
+import math
+import numpy as np
+
+from .._utils import get_array_module
+
+from .base import FourierOperatorBase, CUPY_AVAILABLE, AUTOGRAD_AVAILABLE
+from .interfaces.utils import is_cuda_array
+
+if CUPY_AVAILABLE:
+    import cupy as cp
+
+if AUTOGRAD_AVAILABLE:
+    import torch
+
+
+def get_interpolators_from_fieldmap(
+    fieldmap, readout_time, n_time_segments=6, n_bins=(40, 10), mask=None
+):
+    r"""Create B and C matrices to approximate ``exp(-2j*pi*fieldmap*t)``.
+
+    Here, B has shape ``(n_time_segments, len(t))``
+    and C has shape ``(n_time_segments, *fieldmap.shape)``.
```

Comment:
@paquiteau @chaithyagr that shape makes the matrix quite memory intensive but it is absolutely necessary for R2* + B0 map correction at the same time. During my PhD I just avoided that because it would have been impossible to store on GPU. Should we propose a B0-only less intensive version based on the pysap-mri implementation ?

Reply:
Ok, I had a look to pysap-mri offresonance wrapper. If I understand it correctly, during the operator application it loops over the field coefficients (like my implementation) and it recreate the l-th spatial interpolator for each iteration in the loop:

```
 for l in range(self.num_interpolators):
            y += self.B[..., l] * self.fourier_op.op(
                self.C[l, self.indices] * x,
                *args
            )
```

Memory-wise, this is equal to compute "online" the l-th spatial interpolator as `xp.exp(-1j * tl[l] * field_map)`. I have done [some tests](https://gist.github.com/mcencini/b872bc74f57c52762953f533a57ed9f5) to compare the runtimes of indexing (`self.C[l, self.indices]`) and exp operation for different backends, and these are my results:

```
numpy indexing:
55.3 ms ± 110 μs per loop (mean ± std. dev. of 7 runs, 10 loops each)
numpy exp:
330 ms ± 387 μs per loop (mean ± std. dev. of 7 runs, 1 loop each)
cupy indexing:
475 μs ± 15.6 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
cupy exp:
58.7 μs ± 17.2 μs per loop (mean ± std. dev. of 7 runs, 1 loop each)
torch cpu indexing:
13.9 ms ± 1.68 ms per loop (mean ± std. dev. of 7 runs, 100 loops each)
torch cpu exp:
54.9 ms ± 318 μs per loop (mean ± std. dev. of 7 runs, 10 loops each)
torch cuda indexing:
478 μs ± 5.89 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
torch cuda exp:
31.2 μs ± 15.7 μs per loop (mean ± std. dev. of 7 runs, 1 loop each)
```
Apparently, for CPU (mine is a Intel(R) Xeon(R) Gold 6342 CPU @ 2.80GHz with 24 cores) indexing is faster than exponential, while for GPU (NVIDIA A40) I observe the opposite (exponential is faster). Since storing the whole `(n_time_segments, *shape)` matrix should not be a problem on CPU, I think we could switch to online computation of `C[idx]` if `is_device_array(b0_map)` is True.

Reply:
You did great ! Also you made me realize I wrongly believed it was not possible because of the 2d-histogram to do it in the case of B0+R2* correction. I agree with your suggestions.

On another note, it reminded me of some tests I made with binning precision. The default parameter is (40, 10), how was it chosen exactly ? We could check directly on a typical 3T or 7T field map + R2* map the matrix factorization error depending on the number of bins to suggest the default arguments if not already done, what do you think ?

Reply:
Thank you very much! :D I used the same default as MIRT, but I don't know how this was estimated. Unfortunately I do not dispose of typical 3T / 7T complex field maps right now - maybe we could use the same as your pysap-mri implementation?
```
if (n_bins == "auto"):
            n_bins = 2 * int(np.around(range_w[1] - range_w[0]))
```

Reply:
We can just create an issue to make that study later. I would advise against using the method in my pysap-mri implementation, as it was a rule of thumb that just account for the off-resonance range, while it should also be dependent on the readout length. We can craft a better rule based on a later study

Reply:
Sounds good to me!